### ISSUE TITLE
Can't download text file

### ISSUE THREAD

**Issue by igor-lemon** [parse-community/parse-server#7572]
### New Issue Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).
- [x] I can reproduce the issue with the [latest version of Parse Server](https://github.com/parse-community/parse-server/releases). 

### Issue Description
Can't download a text file. The file opens in a browser, I need to save it manually. 

### Steps to reproduce
Upload some text file. Try to download it by link.

### Expected Outcome
I want to download the file immediantly. 

### Environment

Server
- Parse Server version: `FILL_THIS_OUT`
- Operating system: `FILL_THIS_OUT`
- Local or remote host (AWS, Azure, Google Cloud, Heroku, Digital Ocean, etc): `FILL_THIS_OUT`

Database
- System (MongoDB or Postgres): `FILL_THIS_OUT`
- Database version: `FILL_THIS_OUT`
- Local or remote host (MongoDB Atlas, mLab, AWS, Azure, Google Cloud, etc): `FILL_THIS_OUT`

Client
- SDK (iOS, Android, JavaScript, PHP, Unity, etc): `FILL_THIS_OUT`
- SDK version: `FILL_THIS_OUT`

**Comment by mtrezza** [comment]
Thanks for opening this issue!

I can imagine that people have different expectations, depending on how they use the dashboard. Some may want an image for example to be displayed immediately in the browser, others may want it to be downloaded. Changing this one way or the other may disrupt or improve someone's workflow.

Browsers usually have an option to to download a resource instead of displaying it directly in the browser. For 

For example in Safari on macOS, if you hold the option key while clicking on a link, and the browser decides to download it, regardless of what the server recommends. You can try this out by going [here](https://github.com/parse-community/parse-server/blob/master/package.json) and clicking the `Raw` button in the file header bar.

Conversely, I don't think browsers do have an easily accessible shortcut to display a resource in a browser. So it may make more sense to keep the current behavior of opening resources in the browser by default and give the user the option to use a browser shortcut to download a resource.

**Comment by mtrezza** [comment]
@igor-lemon can this be closed?

**Comment by igor-lemon** [comment]
If you think that it isn't a popular solution you can close the issue and close the PR.

**Comment by mtrezza** [comment]
It changes the behavior, but how can we decide which behavior should be preferred?

**Comment by igor-lemon** [comment]
Maybe my needs are exclusive, but I really don't need to show the file in a browser.
I think the best solution - set custom headers for the file when we save it or add some option `justDownload<bool>`. But the problem that we need to save this info about files to some collection/table.
Or maybe a more easy solution just to set some option when we init the Parse Server, if you like this way I can try to implement it.

**Comment by mtrezza** [comment]
And why don't you use the `download` attribute?

**Comment by igor-lemon** [comment]
> And why don't you use the `download` attribute?

Because it doesn't work :)
Parse generates a not strong path to the file (I use Amazon S3 as file storage)

**Comment by mtrezza** [comment]
What do you mean by "not strong path"?

**Comment by igor-lemon** [comment]
Sorry, not a direct path

**Comment by mtrezza** [comment]
Do you mean the path points directly to S3 and not via Parse Server?

**Comment by igor-lemon** [comment]
Oh, no, I found a problem...
Parse Server doesn't return correct `Content-Type` header, just return  `null` value. 
If I want to set `binary/octet-stream` type
```js
new Parse.File(`${id}.data`, { base64: base64data }, 'binary/octet-stream');
```
It doesn't work...

**Comment by igor-lemon** [comment]
So, I can propose 2 ways:
1) Make `binary/octet-stream` as default content type (if mime couldn't detect type of file)
or
2) Try to Parse-SDK-JS

**Comment by igor-lemon** [comment]
Hi @mtrezza!
Check pls the PR ^

**Comment by mtrezza** [comment]
Maybe let's go one step back just to think this through.

Looking at your code you set the file type:

```js
new Parse.File(`${id}.data`, { base64: base64data }, 'binary/octet-stream');
```

There are two ways of retrieving the file:
- a) via Parse Server
- b) directly from storage

In case of (a), is the file type preserved, so that the issue only occurs for (b)?

**Comment by igor-lemon** [comment]
Check the code in PR. 
Parse Server doesn't check the real content type of file, uses `mime` lib which tries to detect the content type by file extension. I save a file with `*.data` extension and Parse Server return `null` in headers сontent type, and it's not valid.

**Comment by mtrezza** [comment]
Then what's the point of setting the file type when saving the file in your code bit?

**Comment by igor-lemon** [comment]
This question is not for me, you can ask your friends who developed it, why they write that I can set the content type of the file but when I want to download this file parse server return trash in the headers.
Will you merge the PR or not?

**Comment by igor-lemon** [comment]
Yep. I checked the problem just on the Parse Server part, JS SDK works well and content type very important param.
When we create a file we can set content type (an example when we received a Buffer and we know that it's an image and we need to set "image/png" content type). The example of my file:
```bash
$ file --mime-type lic.data 
lic.data: application/octet-stream
```
But when we want to download this file, the parse server just checks the extension from the file name. But if mime lib checks some not standard extension (`*.data`) then returns `null`. I propose a fast fix: leave the same way but if `mime` lib couldn't detect content type by file extension just return 'binary/octet-stream', it provides download file process, of course, the most correct way would be to check on the server-side the real content of the file type and return this value.

**Comment by mtrezza** [comment]
Do you know whether the context type is stored as a parameter of the `Parse.File` on the server? If that was the case, we could use that.

I think the priorities could be:
1. Content type set when uploading a file
2. mimelib detection by file extension

I'm not sure what the side effects are when adding a fallback to `binary/octet-stream` as a third option. That may be an incorrect content type in some cases and maybe it would be better to not specify a content type at all instead of an incorrect one?

**Comment by igor-lemon** [comment]
Yes, `Parse.File` makes a correct file with the correct mime type.
No, just check the code in `src/Routers/FilesRouter.js:71`, parse server doesn't try to detect the content type of an uploaded file, directly detect by file extension and return null for non-standard extensions.

I think that our conversation is going a little in the wrong direction. I do not need help to solve my issues, I just want to convey that the parse-server returns the wrong content type in the headers, which is based only on the file extension from the file name, and not on the real type.

As I wrote above, the best option would be to get the mime info of an uploaded file and return it, but it isn't a little work and I have no desire to do it because we can't even agree on a simple hotfix in 1 line that solves 99% cases. 
I think we are not using our time rationally If you don't want to merge the PR just let me know, I'll use a fork. Thanks! 🙏🏻

**Comment by igor-lemon** [comment]
Maybe you are a little misled, this issue I created just because you request the issue for a pull request. I don’t need to solve any own issues here. :)

**Comment by mtrezza** [comment]
The issue discussion here should justify the PR, not the other way around. That's why opening an issue is a requirement. Maybe we can clarify that in the PR template.

The issue I see is that your suggestion means setting an incorrect content type in some cases if the content type cannot be determined. What is the argument for that and why wouldn't it be better to not specify a content type at all instead of an incorrect one?

**Comment by dblythy** [comment]
Perhaps we could add the option `fileUpload.forceDownload` which adds the following headers: 

`res.set("Content-Disposition", "attachment;filename=somefile.ext");`

**Comment by mtrezza** [comment]
Wouldn't that be something one wants to determine on a per-file basis?

**Comment by dblythy** [comment]
Probably, which I assume would come when we have a beforeGetFile trigger

**Comment by mtrezza** [comment]
That sounds like an exciting idea!

**Comment by dblythy** [comment]
What are your thoughts? I think that the file get triggers have their own considerations and concerns, but it's probably the most versatile approach to this issue

**Comment by mtrezza** [comment]
Yep, agree
